### PR TITLE
Make libkern optimization compatible with purecap kernel

### PR DESCRIPTION
**Pull request by qwattash**

### CODE REVIEW

File: sys/libkern/strlen.c [hunk 1]
```diff
@@ -95,19 +94,34 @@ size_t
 	 * p and (p & ~LONGPTR_MASK) must be equally accessible since
 	 * they always fall in the same memory page, as long as page
 	 * boundaries is integral multiple of word size.
+         *
+         * This is not true for CHERI, so we skip directly to byte
```
Comment (brooksdavis) [inline sys/libkern/strlen.c:98]:
```suggestion
	 *
	 * This is not true for CHERI, so we skip directly to byte
```

File: sys/libkern/strlen.c [hunk 2]
```diff
@@ -95,19 +94,34 @@ size_t
...
+	if (byte_check)
 		/* Check if we have \0 in the first part */
 		for (p = str; p < (const char *)lp; p++)
 			if (*p == '\0')
 				return (p - str);
 
 	/* Scan the rest of the string using word sized operation */
-	for (; ; lp++) {
+#ifdef __CHERI_PURE_CAPABILITY__
+	for (; cheri_getlen(lp) - cheri_getoffset(lp) >= sizeof(long); lp++)
```
Comment (arichardson) [inline sys/libkern/strlen.c:120]: There is also `cheri_bytes_remaining()` in cheric.h. However, that set negative values to zero so might add extra overhead. I think trapping if you pass OOB caps is fine so we probably don't need that.

Maybe we want a  variant of that helper that can return negative values?

Comment (arichardson) [reply sys/libkern/strlen.c:120]: Or an instruction that gives the remaining addressable bytes.

Comment (qwattash) [reply sys/libkern/strlen.c:120]: I'm not sure about adding a `cheri_bytes_remaining` variant just for one use case. I'd leave the explicit subtraction here as we avoid the extra check for zeroing when OOB and feels readable enough.

**Pull request merged**